### ISSUE TITLE
Add the structural similarity (SSIM) index metric as a built-in loss operation

### ISSUE THREAD

**Issue by datbui** [tensorflow/tensorflow#15370]
In many cases existed built-in losses in TensorFlow do not satisfy needs.  We can add [ssim](https://en.wikipedia.org/wiki/Structural_similarity) or (1-ssim) as the loss function into TensorFlow.

There is existed solution provided on [StackOverflow](https://stackoverflow.com/questions/39051451/ssim-ms-ssim-for-tensorflow), but it is better to have the built-in function with fully covered unit tests.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue? Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by datbui** [comment]
Dear tensorflow,

The feature request still actual.


### System information
- **Have I written custom code**: No, i have not
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Mac OS X 10.12.6
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.4.0
- **Python version**: 3.6.2
- **CUDA/cuDNN version**: no cuda installed
- **GPU model and memory**: built-in gpu 
- **Exact command to reproduce**:

**Comment by tensorflowbutler** [comment]
The original poster has replied to this issue after the stat:awaiting response label was applied.

**Comment by drpngx** [comment]
That sounds reasonable. If we'd encourage a submission, then we would need to make sure we have a maintainer.

**Comment by alexlee-gk** [comment]
One thing to note is that you can get different values depending on implementation details, which is not good when reporting numbers in papers. Here is yet another implementation, which I based it off scikit-image's compare_ssim implementation (and the numbers match): https://gist.github.com/alexlee-gk/cbc9bfa6e5be51b53c622684cec0a3f3

**Comment by gabrieldemarmiesse** [comment]
Hi, I converted the numpy version of ms-ssim in tensorflow-contrib to a TensorFlow version. I could add it to tensorflow and maintain it, but I've never contributed to tensorflow before. I've looked at the contributing.md, so the procedure is clear to me. I just don't know in which file to put it. I see there are losses in Tensorflow contrib, but it seems depreciated. Should I still put it in there? Thanks for the help.

Here is the reference implementation that I used, it seems based on the matlab reference implementation of the author of the MS-SSIM.

https://github.com/tensorflow/models/blob/master/research/compression/image_encoder/msssim.py

Please tell me if you're willing to have the MS-SSIM build-in in tensorflow, and if yes, in which file I should put it. I'll do a proper PR then.

**Comment by cancan101** [comment]
Any chance that those functions can support 3D (volumetric) data ?

**Comment by fcaliva** [comment]
@cancan101 have you found a solution to your problem?
